### PR TITLE
feat: Add the auto-coded section in the add-edit-expense form

### PR DESCRIPTION
**Pull request by Z3RO-O**
## Clickup
https://app.clickup.com/t/86cw9uyqc

## Code Coverage
Please add code coverage here

## UI Preview
image

## Summary by CodeRabbit

- **New Features**
  - Introduced visual indicators for auto-coded fields in the expense form, enhancing user feedback.
  - Added conditional display for auto-coded messages in the currency component.

- **Styling Enhancements**
  - New styles for auto-coded elements and sparkle icons, improving visual clarity and aesthetics.

### CODE REVIEW

File: src/app/fyle/add-edit-expense/add-edit-expense.page.ts [hunk 1]
```diff
@@ -5187,4 +5187,8 @@ export class AddEditExpensePage implements OnInit {
 
     await popover.present();
   }
+
+  isDateAutoCoded() {
```
Comment (suyashpatil78) [inline src/app/fyle/add-edit-expense/add-edit-expense.page.ts:5191]: add return type

File: src/app/fyle/add-edit-expense/add-edit-expense.page.html [hunk 2]
```diff
@@ -320,6 +321,10 @@
                       [name]="txnFields.txn_dt?.field_name || 'Date of Spend'"
                     />
                   </div>
+                  <div class="add-edit-expense--auto-coded" *ngIf="isDateAutoCoded()">
```
Comment (suyashpatil78) [inline src/app/fyle/add-edit-expense/add-edit-expense.page.html:324]: We should never use function calls in angular template expressions
https://medium.com/showpad-engineering/why-you-should-never-use-function-calls-in-angular-template-expressions-e1a50f9c0496

**Review by suyashpatil78 — CHANGES_REQUESTED** [review]
few comments and doubts

**Review by suyashpatil78 — APPROVED** [review]

File: src/app/fyle/add-edit-expense/add-edit-expense.page.html [hunk 3]
```diff
@@ -904,6 +922,13 @@
                 [placeholder]="txnFields.vendor_id?.placeholder"
               >
               </app-virtual-select>
+              <div
+                class="add-edit-expense--auto-coded"
+                *ngIf="inpageExtractedData && inpageExtractedData?.vendor_name === fg.controls.vendor_id.value?.display_name"
```
Comment (Chethan-Fyle) [inline src/app/fyle/add-edit-expense/add-edit-expense.page.html:927]: What if both `inpageExtractedData?.vendor_name` and `fg.controls.vendor_id.value?.display_name` are undefined? Check this case and update the condition if need be.

File: src/app/shared/components/fy-currency/fy-currency.component.ts [hunk 4]
```diff
@@ -163,6 +169,26 @@ export class FyCurrencyComponent implements ControlValueAccessor, OnInit, OnChan
...
+    const isAmountAutoCoded = amount === formAmount;
+
+    this.autoCodeMessage =
+      isCurrencyAutoCoded && isAmountAutoCoded
+        ? 'Currency and Amount are auto coded.'
+        : isCurrencyAutoCoded
+        ? 'Currency is auto coded.'
+        : isAmountAutoCoded
+        ? 'Amount is auto coded.'
+        : ''; // Empty string if neither are auto-coded
```
Comment (Chethan-Fyle) [inline src/app/shared/components/fy-currency/fy-currency.component.ts:191]: Use `if..else..`.

File: src/app/fyle/add-edit-expense/add-edit-expense.page.html [hunk 5]
```diff
@@ -488,6 +496,16 @@
                       [placeholder]="txnFields.org_category_id.placeholder"
                     >
                     </app-virtual-select>
+                    <div
+                      class="add-edit-expense--auto-coded"
+                      *ngIf="inpageExtractedData && inpageExtractedData?.category === fg.controls.category.value?.displayName"
```
Comment (Chethan-Fyle) [inline src/app/fyle/add-edit-expense/add-edit-expense.page.html:501]: Make sure both `inpageExtractedData?.category` and `fg.controls.category.value?.displayName` are not undefined.

File: src/app/fyle/add-edit-expense/add-edit-expense.page.html [hunk 6]
```diff
@@ -320,6 +321,13 @@
                       [name]="txnFields.txn_dt?.field_name || 'Date of Spend'"
                     />
                   </div>
+                  <div
+                    class="add-edit-expense--auto-coded"
+                    *ngIf="this.inpageExtractedData && this.inpageExtractedData?.date === this.fg.controls.dateOfSpend.value"
```
Comment (Chethan-Fyle) [inline src/app/fyle/add-edit-expense/add-edit-expense.page.html:326]: Same as below.

**Review by Chethan-Fyle — COMMENTED** [review]
Few comments.

File: src/assets/svg/stars-filled.svg [hunk 7]
```diff
@@ -1,4 +1,4 @@
 <svg width="24" height="24" viewBox="0 0 24 24" fill="none" xmlns="http://www.w3.org/2000/svg">
```
Comment (Dimple16) [inline src/assets/svg/stars-filled.svg:1]: We shouldn't have the height and width. You can remove them and use CSS to give the icon height and width

Comment (Dimple16) [reply src/assets/svg/stars-filled.svg:1]: Don't we already have the icon in the other PR?

File: src/app/shared/icon/icon.module.ts [hunk 8]
```diff
@@ -115,6 +115,7 @@ export class IconModule {
     'sparkle.svg',
     'arrow-tail-right.svg',
     'stars-filled.svg',
+    'stars-color-filled.svg',
```
Comment (Dimple16) [inline src/app/shared/icon/icon.module.ts:118]: are both the above icons same?

**Review by Dimple16 — CHANGES_REQUESTED** [review]
added comments

File: src/app/shared/components/fy-currency/fy-currency.component.ts [hunk 9]
```diff
@@ -163,6 +169,22 @@ export class FyCurrencyComponent implements ControlValueAccessor, OnInit, OnChan
...
+    const formCurrency = this.fg?.value.currency;
+    const formAmount = this.fg?.value.amount;
+
+    const isCurrencyAutoCoded = currency === formCurrency;
+    const isAmountAutoCoded = amount === formAmount;
+
+    if (isCurrencyAutoCoded && isAmountAutoCoded) this.autoCodeMessage = 'Currency and Amount are auto coded.';
+    else if (isCurrencyAutoCoded) this.autoCodeMessage = 'Currency is auto coded.';
+    else if (isAmountAutoCoded) this.autoCodeMessage = 'Amount is auto coded.';
+    else this.autoCodeMessage = '';
```
Comment (Chethan-Fyle) [inline src/app/shared/components/fy-currency/fy-currency.component.ts:187]: Use braces even if it is a single line. It is the convention we use in our codebase.
```
if (...) {
...
} else if (...) {
...
} else if (...) {
...
}
```

**Review by Chethan-Fyle — APPROVED** [review]
One comment. Please fix and proceed.

**Comment by Z3RO-O** [comment]
lint workflow is failing because of use of any and no return type in functions

**Review by Dimple16 — APPROVED** [review]

**Pull request closed without merging**